Remove hard coded service ips for coredns and kubernetes service

Ebaneck opened the pull request:
**Component**:

<!-- E.g. 'salt', 'containers', 'kubernetes', 'build', 'tests'... -->
'salt', 'kubernetes', 'networking'

**Context**: 

See #780 

**Summary**:

- We need to remove the hard-coded values related to the Coredns cluster IP and Kubernetes cluster IP.
For this, we have written a salt execution module to pick the following:
- The `kubernetes ClusterIP` address which is agreed to be the 2nd address in the service_network.
- The `coredns ClusterIP` address which is the 10th address in the service_network
 For this to work properly, we need to define a proper service network within the pillars

How to test?

```
[root@bootstrap vagrant]# kubectl get pods,svc --namespace=kube-system | grep dns

pod/coredns-69bb6dc74d-48glm                 0/1       ContainerCreating   0          1m
pod/coredns-69bb6dc74d-4tqtj                 0/1       ContainerCreating   0          1m
service/coredns                                        ClusterIP   10.96.0.10       <none>        53/UDP,53/TCP,9153/TCP       1m
```
**Acceptance criteria**: 

- Working metalk8s cluster after bootstrap
- CI build should still pass.

Needed:

Once this ticket is closed, I plan to open another to update the documentation.
---

<!-- Declare one or more issues to close once this PR gets merged -->

Closes: #780

bert-e commented:
# Hello ebaneck,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @teddyandrieux

* @monpote

* @jbertran

* @sayf-eddine-scality

* @nicolast

* @cmonfort

* @slaperche-scality

* @thomasdanan

* @zempashi

* @lucieleonard

* @gdemonet

* @carlito-scality

* @nootal

* @alexandre-allard-scality

Ebaneck commented:
This is a WIP, I am going to ask for reviewers when a fully functional draft is ready.

slaperche-scality commented:
You should target your PR on 2.0, not 2.4 :wink:

Ebaneck committed:
salt: Add a salt module for network related computations.

We need to be able to compute the cluster ip for CoreDNS and
Kubernetes.
For this, we have written a salt execution module to perform this computations.

Ebaneck committed:
salt: Replace hardcoded values for CoreDNS and Kubernetes cluster IP

In this commit, we have done the following:

- Remove hard coded values in default.yaml
- Remove hard coded values in sls files while replacing them with the computed
values from the network module

bert-e commented:
# Branches have diverged



This pull request's source branch `bugfix/780-Remove-hard-coded-service-IPS` has diverged from
`development/2.4` by more than 50 commits.

To avoid any integration risks, please re-synchronize them using one of the
following solutions:

* Merge `origin/development/2.4` into `bugfix/780-Remove-hard-coded-service-IPS`
* Rebase `bugfix/780-Remove-hard-coded-service-IPS` onto `origin/development/2.4`

Note: If you choose to rebase, you may have to ask me to rebuild
integration branches using the `reset` command.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/780-Remove-hard-coded-service-IPS` into
`development/2.0`
* [w/2.1/bugfix/780-Remove-hard-coded-service-IPS](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/bugfix/780-Remove-hard-coded-service-IPS) will be merged into `development/2.1` 
* [w/2.2/bugfix/780-Remove-hard-coded-service-IPS](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/bugfix/780-Remove-hard-coded-service-IPS) will be merged into `development/2.2` 
* [w/2.3/bugfix/780-Remove-hard-coded-service-IPS](https://github.com/scality/metalk8s/compare/development/2.3...w/2.3/bugfix/780-Remove-hard-coded-service-IPS) will be merged into `development/2.3` 
* [w/2.4/bugfix/780-Remove-hard-coded-service-IPS](https://github.com/scality/metalk8s/compare/development/2.4...w/2.4/bugfix/780-Remove-hard-coded-service-IPS) will be merged into `development/2.4` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @teddyandrieux

* @monpote

* @jbertran

* @sayf-eddine-scality

* @nicolast

* @cmonfort

* @slaperche-scality

* @thomasdanan

* @zempashi

* @lucieleonard

* @gdemonet

* @carlito-scality

* @nootal

* @alexandre-allard-scality

bert-e commented:
# History mismatch



Merge commit #3f6bc98277aacf05510c98d4895c59ad2929946f on the integration branch
`w/2.1/bugfix/780-Remove-hard-coded-service-IPS` is merging a branch which is neither the current
branch `bugfix/780-Remove-hard-coded-service-IPS` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `bugfix/780-Remove-hard-coded-service-IPS` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

slaperche-scality commented on salt/_modules/metalk8s_network.py:
```diff
@@ -0,0 +1,45 @@
+'''Metalk8s network related utilities.'''
+
+import ipaddress
```
`New in version 3.3.`

I thought we were stuck to Python 2 for Salt (see this [reply](https://github.com/scality/metalk8s/issues/780#issuecomment-528441446)) ?
Or are we somehow magically pulling the backport to 2.X from pip or something?

Ebaneck replied:
Looks like the module is available by default in python2.

Here: 
```
[root@bootstrap vagrant]# python
Python 2.7.5 (default, Oct 30 2018, 23:45:53) 
[GCC 4.8.5 20150623 (Red Hat 4.8.5-36)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import ipaddress
>>> 
```

gdemonet replied:
No it's not _available by default_, but actually comes from a backport as @slaperche-scality suggested: https://pypi.org/project/ipaddress/

However, this module is `Required-by` the `docker` client, which you probably have installed on your machine. We could add the dependency to the project (as we do with `m2crypto` for instance), or move back to relying on a more manual approach (using `__utils__` for instance).

gdemonet replied:
Now, Salt uses this as well, relying on their `salt._compat` layer: [see this import in the `salt.utils.network` module](https://github.com/saltstack/salt/blob/develop/salt/utils/network.py#L30). Let's do the same, should be sufficient:

```suggestion
from salt._compat import ipaddress
```

slaperche-scality commented on salt/_modules/metalk8s_network.py:
```diff
@@ -0,0 +1,45 @@
+
+
+def pick_nth_ip(service_network, index):
+
```
this empty line should be removed.

slaperche-scality commented on salt/_modules/metalk8s_network.py:
```diff
@@ -0,0 +1,45 @@
+     service network
+
+    :returns (str):
+        An index param of 10 returns => coredns cluster ip
```
What is `an index param`?
Also, what the heck does 2 and 10 means in this context?

We should have named constant or something, rather than relying on those some magical values?

And/or have two public functions `get_service_ip` and `get_dns_ip` and that will call that one.

Ebaneck replied:
I think we had clearly defined what an index `param` signifies. no??

Also FYI:
```
 :param index (int): Kubernetes cluster IP address is set to 2nd address in
     service_network.
     coredns cluster IP address is set to the 10th address in the
     service network
```
As such index=10 => we are computing the ip address for cluster DNS

I do not think it would be any better creating two functions that perform the same actions and so this is the reason we have one generic function doing the same action.

Would mark this as resolved unless another thinks otherwise

Ebaneck replied:
Maybe it should read `An index value of 10 returns a CoreDNS cluster IP`

slaperche-scality commented on salt/_modules/metalk8s_network.py:
```diff
@@ -0,0 +1,45 @@
+def pick_nth_ip(service_network, index):
+
+    '''
+    Pick the coredns and kubernetes cluster ip address from service network.
```
use proper capitalization: `CoreDNS`, `Kubernetes`, `IP address`.

Ebaneck replied:
ok

slaperche-scality commented on salt/_modules/metalk8s_network.py:
```diff
@@ -0,0 +1,45 @@
+    log.info('Picking the nth address from service network')
+
+    network = ipaddress.IPv4Network(unicode(service_network))
+    for i, ip in enumerate(network):
```
could be simplified as:

```python
ip = next(itertools.islice(network.hosts(), index, None), None)
if ip is None:
    raise CommandExecutionError(
        'Could not obtain an IP in the network range {}'.format(
            service_network
        )
    )
return str(ip)
```

Ebaneck replied:
- Adds one more import for `itertools` python module
- `network.hosts()` looks to return strictly the usable set of IP addresses meaning passing an index of 2 returns =>  a CoreDNS IP of 192.168.1.3.
This is not exactly what we want even though we can make a few changes else where to match this.

I prefer to stick to what is currently written. :)

gdemonet replied:
I'm not convinced it's a "simplification", I feel it's quite harder to read :/

However, since using `ipaddress` may not be the chosen strategy, it may not make sense to bother further with this...

slaperche-scality commented on salt/metalk8s/defaults.yaml:
```diff
@@ -113,7 +113,6 @@ kube_api:
   cert:
     server_signing_policy: kube_apiserver_server_policy
     client_signing_policy: kube_apiserver_client_policy
```
Could be squashed with the previous commit, no?

Ebaneck replied:
It could be.

But this still looks okay to me.
one commit => add module + add module path to salt_tree.
one commit => use module in sls files while replacing variables were need be.
one commit => remove the hard coded service IP addresses within the `defaults.yaml`.

Looks ok to me

gdemonet replied:
@Ebaneck I agree with Sylvain, one commit to introduce functionality (the module), one to use the functionality (replace values where need be AND remove now-unused values). If you keep the last commit as you suggest, its value is roughly the same as a `fix typo` commit, which we don't like much ;)

bert-e commented:
# History mismatch



Merge commit #3f8af280635bcbe6dbad166ea2bab5907cd70ab7 on the integration branch
`w/2.1/bugfix/780-Remove-hard-coded-service-IPS` is merging a branch which is neither the current
branch `bugfix/780-Remove-hard-coded-service-IPS` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `bugfix/780-Remove-hard-coded-service-IPS` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

gdemonet commented on salt/_modules/metalk8s_network.py:
```diff
@@ -0,0 +1,44 @@
+    return __virtualname__
+
+
+def pick_nth_ip(service_network, index):
```
As @slaperche-scality suggested, it makes sense to define this as a _private_ helper function, and define specific methods for `get_kubernetes_service_ip` and `get_cluster_dns_ip`.

The rationale is simple: even if `pick_nth_ip` is an explicit name w.r.t what it does, it doesn't tell us **why** we want to pick the 10th for DNS, nor the first for the `kubernetes` service.

So, this module could instead look like:

```
# imports and other init stuff...

def _pick_nth_ip(network_range, n):
    """Pick the nth IP address from a network range, provided in CIDR notation."""
   # Same code...
   return ip

def get_cluster_dns_ip():
    """Return the CoreDNS cluster IP based on pillar configuration.

    This IP is arbitrarily selected as the tenth IP from the Service network range.
    """
    return _pick_nth_ip(__pillar__['networks.service'], 10)


def get_kubernetes_service_ip():
    """Return the Kubernetes service cluster IP based on pillar configuration.

    This IP is arbitrarily selected as the second IP from the Service network range.
    """
    return _pick_nth_ip(__pillar__['networks.service'], 2)
```

It yields that, in SLS files, no one will wonder "_why do we pick these values?!_", and instead will refer to these methods for understanding. You could have achieved the same result by adding a comment in those SLS files, but since these IPs are used several times, you gain the benefit of de-duplication.

gdemonet commented on salt/_modules/metalk8s_network.py:
```diff
@@ -0,0 +1,61 @@
+    return __virtualname__
+
+
+def get_service_ip(service_network):
```
You can factorize these with a helper method, called `_pick_nth_ip` ;)

Ebaneck commented:
/approve

bert-e commented:
# History mismatch



Merge commit #3f8af280635bcbe6dbad166ea2bab5907cd70ab7 on the integration branch
`w/2.1/bugfix/780-Remove-hard-coded-service-IPS` is merging a branch which is neither the current
branch `bugfix/780-Remove-hard-coded-service-IPS` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `bugfix/780-Remove-hard-coded-service-IPS` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **approve**

Ebaneck commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **approve**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/780-Remove-hard-coded-service-IPS` into
`development/2.0`
* [w/2.1/bugfix/780-Remove-hard-coded-service-IPS](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/bugfix/780-Remove-hard-coded-service-IPS) will be merged into `development/2.1` 
* [w/2.2/bugfix/780-Remove-hard-coded-service-IPS](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/bugfix/780-Remove-hard-coded-service-IPS) will be merged into `development/2.2` 
* [w/2.3/bugfix/780-Remove-hard-coded-service-IPS](https://github.com/scality/metalk8s/compare/development/2.3...w/2.3/bugfix/780-Remove-hard-coded-service-IPS) will be merged into `development/2.3` 
* [w/2.4/bugfix/780-Remove-hard-coded-service-IPS](https://github.com/scality/metalk8s/compare/development/2.4...w/2.4/bugfix/780-Remove-hard-coded-service-IPS) will be merged into `development/2.4` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @teddyandrieux

* @monpote

* @jbertran

* @sayf-eddine-scality

* @nicolast

* @cmonfort

* @slaperche-scality

* @thomasdanan

* @zempashi

* @lucieleonard

* @gdemonet

* @carlito-scality

* @nootal

* @alexandre-allard-scality





The following reviewers are expecting changes from the author, or must review again:

* @slaperche-scality

* @gdemonet







*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @teddyandrieux

* @monpote

* @jbertran

* @sayf-eddine-scality

* @nicolast

* @cmonfort

* @slaperche-scality

* @thomasdanan

* @zempashi

* @lucieleonard

* @gdemonet

* @carlito-scality

* @nootal

* @alexandre-allard-scality





The following reviewers are expecting changes from the author, or must review again:

* @gdemonet







*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/38725a583797ca1d49fe7e5979c195621c758f8d) did not succeed in branch bugfix/780-Remove-hard-coded-service-IPS.




*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/745b138f5ebd8435a12549e6c774d66ce2caaad3) did not succeed in branch w/2.2/bugfix/780-Remove-hard-coded-service-IPS.




*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/0c18eb079d9ba4b4629daa5640f61407155e2f1f) did not succeed in branch w/2.1/bugfix/780-Remove-hard-coded-service-IPS.




*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/38725a583797ca1d49fe7e5979c195621c758f8d) did not succeed in branch bugfix/780-Remove-hard-coded-service-IPS.




*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/175fc1afad4a6d2698bdfd7349ba57b625e4a8c8) did not succeed in branch w/2.4/bugfix/780-Remove-hard-coded-service-IPS.




*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/0c18eb079d9ba4b4629daa5640f61407155e2f1f) did not succeed in branch w/2.1/bugfix/780-Remove-hard-coded-service-IPS.




*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/38725a583797ca1d49fe7e5979c195621c758f8d) did not succeed in branch bugfix/780-Remove-hard-coded-service-IPS.




*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/0c18eb079d9ba4b4629daa5640f61407155e2f1f) did not succeed in branch w/2.1/bugfix/780-Remove-hard-coded-service-IPS.




*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/38725a583797ca1d49fe7e5979c195621c758f8d) did not succeed in branch bugfix/780-Remove-hard-coded-service-IPS.




*The following options are set:* **approve**

bert-e commented:
# History mismatch



Merge commit #064b9a93bf06aa110e1526a55b0d7e00002436b3 on the integration branch
`w/2.1/bugfix/780-Remove-hard-coded-service-IPS` is merging a branch which is neither the current
branch `bugfix/780-Remove-hard-coded-service-IPS` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `bugfix/780-Remove-hard-coded-service-IPS` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **approve**

Ebaneck commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **approve**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/780-Remove-hard-coded-service-IPS` into
`development/2.0`
* [w/2.1/bugfix/780-Remove-hard-coded-service-IPS](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/bugfix/780-Remove-hard-coded-service-IPS) will be merged into `development/2.1` 
* [w/2.2/bugfix/780-Remove-hard-coded-service-IPS](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/bugfix/780-Remove-hard-coded-service-IPS) will be merged into `development/2.2` 
* [w/2.3/bugfix/780-Remove-hard-coded-service-IPS](https://github.com/scality/metalk8s/compare/development/2.3...w/2.3/bugfix/780-Remove-hard-coded-service-IPS) will be merged into `development/2.3` 
* [w/2.4/bugfix/780-Remove-hard-coded-service-IPS](https://github.com/scality/metalk8s/compare/development/2.4...w/2.4/bugfix/780-Remove-hard-coded-service-IPS) will be merged into `development/2.4` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @teddyandrieux

* @monpote

* @jbertran

* @sayf-eddine-scality

* @nicolast

* @cmonfort

* @slaperche-scality

* @thomasdanan

* @zempashi

* @lucieleonard

* @gdemonet

* @carlito-scality

* @nootal

* @alexandre-allard-scality










*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/b27190e39deb22e3ced96514ffd9079bce29b341) did not succeed in branch w/2.4/bugfix/780-Remove-hard-coded-service-IPS.




*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.0`

* :heavy_check_mark: `development/2.1`

* :heavy_check_mark: `development/2.2`

* :heavy_check_mark: `development/2.3`

* :heavy_check_mark: `development/2.4`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.0`

* :heavy_check_mark: `development/2.1`

* :heavy_check_mark: `development/2.2`

* :heavy_check_mark: `development/2.3`

* :heavy_check_mark: `development/2.4`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`



Please check the status of the associated issue None.

Goodbye ebaneck.
